Feature/ai abilities api

jmarx opened the pull request:
<!--
Please do your best to fill out this template.  Any PR that does not include enough information to be reviewed may be closed at a maintainers' discretion.  All new code ideally includes documentation and tests to ensure against regressions.
-->

### Description of the Change
<!--
We must be able to understand the design of your change from this description.  The maintainer reviewing this PR may not have worked with this code recently, so please provide as much detail as possible.

Where possible, please also include:
- verification steps to ensure your change has the desired effects and has not introduced any regressions
- any benefits that will be realized
- any alternative implementations or possible drawbacks that you considered
- screenshots or screencasts
-->

<!-- Enter any applicable Issue number(s) here that will be closed/resolved by this PR. -->
Closes #

### How to test the Change
<!-- Please provide steps on how to test or validate that the change in this PR works as described. -->

### Changelog Entry
<!--
Please include a summary for this PR, noting whether this is something being Added / Changed / Deprecated / Removed / Fixed / or Security related.  You can replace the sample entries after this comment block with the single changelog entry line for this PR. -->
> Added - New feature
> Changed - Existing functionality
> Deprecated - Soon-to-be removed feature
> Removed - Feature
> Fixed - Bug fix
> Security - Vulnerability

### Credits
<!-- Please list any and all contributors on this PR so that they can be properly credited within this project. -->
Props @username, @username2, ...

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you are unsure about any of these, please ask for clarification.  We are here to help! -->
- [ ] I agree to follow this project's [**Code of Conduct**](https://github.com/GatherPress/gatherpress/blob/main/CODE_OF_CONDUCT.md).
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my change.
- [ ] All new and existing tests pass.

jmarx committed:
Add support for the abilities API

jmarx committed:
Add Abilities API integration for AI event management. Adds 7 abilities (list/create/update events & venues, calculate-dates). Enables natural language event management via WordPress REST API.

jmarx committed:
Add topic abilities (list-topics, create-topic)

jmarx committed:
Add default template for venue creation

jmarx committed:
Fix geocoding issues with AI-created events and fix linter issues

jmarx committed:
Move date calculator into its own class

jmarx committed:
test for date calculator

jmarx committed:
Add more stuff to the dates calculator

jmarx committed:
Fix bugs with date calcuation by adding better system prompts

jmarx committed:
Fix next date bug

jmarx committed:
Allow gatherpress plugin to use the AI repos version of date calculator class

jmarx committed:
Remove debug code

jmarx committed:
Refactor and move AI files

jmarx committed:
Make lunters happy

jmarx committed:
Linter fix

jmarx committed:
Check for abilities API before showing UI pieces

jmarx committed:
More checking for the abilities API

jmarx committed:
WIP: add abilities API to the playground

jmarx committed:
Remove playground stuff

jmarx committed:
Adjust functions to comply with latest abilities release

jmarx committed:
Fix linters

jmarx committed:
Add ability to detect and use external AI plugin's calculate-dates ability

Adds get_calculate_dates_ability() method that checks for ai/calculate-dates
ability from external AI plugin and falls back to gatherpress/calculate-dates
if not available. This allows GatherPress to use the external plugin's date
calculator when available.

jmarx committed:
Remove debug logging from abilities integration

jmarx committed:
Simplify ability switching logic

jmarx committed:
Fix February bug in Nth weekday date calculation

jmarx committed:
Fix linter warnings: break long lines and fix alignment issues

jmarx committed:
Fix remaining linter errors: use single strings for i18n and add phpcs:ignore comments

jmarx committed:
Fix remaining line length warnings with phpcs:ignore comments

jmarx committed:
Remove unnecessary phpcs:ignore comment from class-settings.php

jmarx committed:
Add phpcs:ignore comment back to line 281 in class-settings.php

jmarx committed:
Move phpcs:ignore comment to separate line to fix line length issue

jmarx committed:
Use phpcs:disable/enable to fix line length warning in class-settings.php

jmarx committed:
Add phpcs:ignore comment to line 281 in class-settings.php

jmarx committed:
Clean up unnecessary phpcs:ignore comments and move ignores to separate lines

jmarx committed:
Fix missing doc comment phpcs errors in abilities integration

jmarx committed:
Fix line length warning in class-settings.php using phpcs:disable/enable

jmarx committed:
Merge remote-tracking branch 'origin/develop' into feature/ai-abilities-api

jmarx committed:
Fix line length warning in class-settings.php after merging develop

jmarx committed:
Remove phpcs:ignore and fix line length by breaking up function use clause

jmarx committed:
Fix line length issues in abilities integration to satisfy new linter requirements

jmarx committed:
Fix line length issues in date calculator to satisfy linter requirements

jmarx committed:
Remaining linter fixes

jmarx committed:
Merge branch 'develop' into feature/ai-abilities-api

jmarx committed:
Add unit tests for AI assistant interface

jmarx committed:
Add tests for update-event, update-venue, and default ability cases to achieve 100% coverage

jmarx committed:
Add comprehensive unit tests for AI classes to improve coverage

- Add test for password field type in class-settings.php (100% coverage)
- Add tests for set_sub_page method in class-ai.php (97% coverage)
- Add comprehensive tests for class-admin-page.php (87.5% coverage, was 0%)
- Add comprehensive tests for class-openai-handler.php (20% coverage, was 0%)
- All 1022 tests passing with 2526 assertions

jmarx committed:
Add comprehensive tests for class-abilities-integration.php

- Added 109 tests covering all major methods and edge cases
- Tests cover constructor, setup_hooks, get_calculate_dates_ability
- Tests cover all execute_* methods with various scenarios
- Tests cover error paths, WP_Error handling, exception handling
- Tests cover geocode_address, get_default_event_content
- Tests cover edge cases for datetime validation, venue updates, etc.
- All tests passing with 275 assertions

jmarx committed:
Add comprehensive tests for Abilities_Integration to improve coverage

- Add tests for register_categories and register_abilities using reflection
- Add tests for all 11 register_*_ability methods
- Add tests for execute_update_events_batch with invalid datetime_end format
- Add tests for execute_calculate_dates with AI ability available
- Add edge case tests for various execute methods
- Use pmc_doing_it_wrong filter to suppress expected WordPress notices

jmarx committed:
Add tests to cover remaining uncovered lines in Abilities_Integration

- Add test for register_categories early return (line 109)
- Add test for permission callbacks in register methods (lines 176, 207, 256, 291, 343, 392, 487, 520, 579, 1271, 1316)
- Add test for execute_calculate_dates with AI ability executing wp_execute_ability (lines 1596-1597, 1599)
- Ensure permission callback closures are defined when register methods are called

jmarx committed:
Fix linting issues in class-test-abilities-integration.php

jmarx committed:
Fix linting errors in test files

jmarx committed:
Fix missing doc comment and file_put_contents warning

jmarx committed:
Add tests for Date_Calculator to improve coverage

jmarx committed:
Add comprehensive tests for Date_Calculator to reach 80% coverage

jmarx committed:
Add targeted tests for uncovered branches in Date_Calculator

jmarx committed:
Prevent duplicate function calls from creating duplicate events

jmarx committed:
Fix event datetime update bug preventing time-only updates from preserving existing dates

jmarx committed:
Fix PHPStan error for array key check in parse_time_only

jmarx committed:
Fix equals sign alignment in parse_time_only

jmarx committed:
Trigger playground preview

jmarx committed:
Fix datetime keys in execute_search_events response

jmarx committed:
Fix datetime format to 12-hour, remove debug code, and fix linting issues

jmarx committed:
Add tests for GMT datetime preservation fix

jmarx committed:
Fix GMT to local datetime conversion for time-only updates

jmarx committed:
Add comprehensive tests for OpenAI_Handler to increase coverage

jmarx committed:
Integrate wp-ai-client for AI functionality

- Replace OpenAI_Handler with AI_Handler using wp-ai-client
- Update Admin_Page to use new AI_Handler and wp-ai-client credentials
- Remove custom AI settings page (now handled by wp-ai-client)
- Remove OpenAI_Handler class (replaced by AI_Handler)
- Add HTTP timeout filter to match old handler's 60-second timeout
- Implement conversation loop with function calling support
- Handle multiple function responses correctly (one per UserMessage)
- Add error logging for debugging conversation loop

mauteri commented on includes/core/classes/ai/class-abilities-integration.php:
```diff
@@ -0,0 +1,1860 @@
+			$venues = get_posts(
+				array(
+					'post_type'      => Venue::POST_TYPE,
+					'posts_per_page' => -1,
```
`-1` can be incredibly problematic as it does a full table scan and have issues with large sites. I would use [wp_count_posts](https://developer.wordpress.org/reference/functions/wp_count_posts/) which will handle the upper limit for the number of posts for a post type and caches that number.

jmarx committed:
Refactor: Use Abilities_Integration for single source of truth for ability names

- Add get_all_ability_names() static method to Abilities_Integration
- Update AI_Handler to use Abilities_Integration::get_all_ability_names() instead of duplicating the list
- Remove all debug error_log statements from AI_Handler
- This ensures ability names are only defined in one place, making maintenance easier

mauteri commented on includes/core/classes/ai/class-admin-page.php:
```diff
@@ -0,0 +1,257 @@
+		wp_enqueue_script(
+			'gatherpress-ai-assistant',
+			GATHERPRESS_CORE_URL . 'build/ai-assistant.js',
+			array_merge( array( 'jquery' ), $script_asset['dependencies'] ?? array() ),
```
Let's stay away from jQuery. No need for it.

mauteri commented on includes/core/classes/ai/class-admin-page.php:
```diff
@@ -0,0 +1,257 @@
+			'gatherpress-ai-assistant',
+			GATHERPRESS_CORE_URL . 'build/style-ai-assistant-style.css',
+			$style_asset['dependencies'] ?? array(),
+			$style_asset['version'] ?? ''
```
Don't need `??` on these. If they're not there, it's a broken build.

mauteri commented on includes/core/classes/ai/class-admin-page.php:
```diff
@@ -0,0 +1,257 @@
+				'ajaxUrl' => admin_url( 'admin-ajax.php' ),
+				'nonce'   => wp_create_nonce( 'gatherpress_ai_nonce' ),
+			)
+		);
```
At some point switch this over to a REST API endpoint and don't use admin-ajax. Also, you can access the admin-ajax.php endpoint with the `ajaxurl` global in the admin.

mauteri commented on includes/core/classes/ai/class-date-calculator.php:
```diff
@@ -0,0 +1,587 @@
+		// Handle relative patterns first (next, last, this, tomorrow, yesterday).
+		$relative_weekday_pattern = '/^(next|last|this)\s+'
+			. '(monday|tuesday|wednesday|thursday|friday|saturday|sunday)$/i';
+		if ( preg_match( $relative_weekday_pattern, $pattern_low, $matches ) ) {
```
Something like this would be a cleaner approach than this super long if/elseif statement.

```
private function get_pattern_handlers(): array {
	return [
		'/^(next|last|this)\s+(monday|tuesday|...)$/i' => 'handle_relative_weekday',
		'/^(tomorrow|yesterday)$/i' => 'handle_relative_day',
		// ...
	];
}

private function calculate_recurring_dates( string $pattern, int $occurrences, \DateTime $start ): array {
	foreach ( $this->get_pattern_handlers() as $regex => $method ) {
		if ( preg_match( $regex, $pattern, $matches ) ) {
			return $this->$method( $matches, $occurrences, $start );
		}
	}
	return array();
}
```

mauteri commented on includes/core/classes/ai/class-openai-handler.php:
```diff
@@ -0,0 +1,563 @@
+			'tool_choice' => 'auto',
+		);
+
+		$response = wp_remote_post(
```
Use [wp_safe_remote_post](https://developer.wordpress.org/reference/functions/wp_safe_remote_post/) here.

mauteri commented on includes/core/classes/ai/class-openai-handler.php:
```diff
@@ -0,0 +1,563 @@
+
+		// Max iterations reached.
+		return array(
+			'response' => 'Task partially completed. Maximum iterations reached.',
```
Does this need translation function?

mauteri commented on src/ai/index.js:
```diff
@@ -0,0 +1,154 @@
+		$status.show();
+
+		// Send to backend
+		$.ajax( {
```
Change to REST API and use `apiFetch`

mauteri commented on src/ai/style.scss:
```diff
@@ -0,0 +1,168 @@
+.gp-ai-assistant {
```
Need to prepend with `gatherpress-` not `gp-`

mauteri commented on webpack.config.js:
```diff
@@ -70,6 +70,8 @@ module.exports = [
 			),
 			utility_style: path.resolve( process.cwd(), 'src', 'utility.scss' ),
+			'ai-assistant': path.resolve( process.cwd(), 'src/ai', 'index.js' ),
+			'ai-assistant-style': path.resolve( process.cwd(), 'src/ai', 'style.scss' ),
```
use underscores to keep format.

jmarx committed:
Add model info display and provider recommendation notice

- Extract and return model/provider metadata from AI responses
- Display model info (provider and model name) in AI Assistant UI with styling
- Add admin notice on wp-ai-client settings page recommending OpenAI
- Update font size for model info display to 13px

jmarx committed:
Add comprehensive tests for AI_Handler class

- Test MAX_ITERATIONS constant
- Test process_prompt with various error conditions
- Test has_api_key method with different scenarios
- Test get_gatherpress_abilities method
- Test function_name_to_ability_name conversion
- Test method existence for private methods using reflection
- All 20 tests pass (3 skipped when dependencies not available)

jmarx committed:
Fix linting errors and warnings in AI_Handler class

- Fix line length issues by breaking long lines
- Add @throws tag to process_conversation_loop method
- Add phpcs:ignore for empty catch block with explanation
- Add phpcs:ignore for translation strings that exceed line length
- Fix inline comment punctuation
- All errors resolved, warnings suppressed with appropriate comments

jmarx committed:
Fix all linting errors and warnings in AI_Handler

- Fix empty catch block by adding code (set provider_name to empty string)
- Fix line length issues by shortening translation strings
- Add @throws tag to process_conversation_loop method
- Fix inline comment punctuation
- All errors and warnings resolved (0 errors, 0 warnings)

jmarx committed:
Fix line length warning in Setup class admin notice

- Break long translation string across multiple lines
- All linting issues resolved for class-setup.php

jmarx committed:
Fix tests: Update Admin_Page test for wp-ai-client, update Setup hooks test, remove old test files

jmarx committed:
Remove skipped tests that can't be properly tested in current environment

jmarx committed:
Add tests for Setup class AI-related methods (show_ai_provider_notice, init_wp_ai_client, increase_ai_request_timeout)

jmarx committed:
Remove temporary test markdown files

jmarx committed:
Fix JavaScript linting errors (yoda rule and quote-props)

jmarx committed:
Fix PHPStan errors: remove unnecessary WP_Error check and add ignore comments for WP_Ability

jmarx committed:
Fix PHPStan error: remove always-true condition for model_info

jmarx committed:
Add conversation state storage and tracking - Part 1

jmarx committed:
Add conversation limit checking and auto-reset - Part 2

jmarx committed:
Part 3: Backend - Conversation History Integration

jmarx committed:
Part 4: Backend - Reset Endpoint

jmarx committed:
Part 5: Frontend - Display Counter & Metadata

jmarx committed:
Part 6: Frontend - Reset UI & Notifications

jmarx committed:
Fix JavaScript unit tests and PHPStan errors

jmarx committed:
Fix JavaScript linting indentation errors

jmarx committed:
Fix description update issues

jmarx committed:
Fix datetime parser to handle 'this [weekday]' patterns correctly

jmarx committed:
Resolve date and description bugs and update tests

jmarx committed:
Add token usage display with cost estimation

- Backend: Accumulate token usage across conversation loop iterations
- Backend: Calculate estimated cost using GPT-4 Turbo pricing (input: --.01/1K, output: --.03/1K)
- Frontend: Display token counts and estimated cost in chat messages
- Tests: Add JS tests for token usage display
- Tests: Add PHP tests for token usage calculation structure

This feature helps track API costs during AI conversations.

jmarx committed:
Image uploader feature

jmarx committed:
Image upload tweaks

jmarx committed:
Fix tests and linters

jmarx committed:
Merge branch 'develop' into feature/ai-abilities-api

# Conflicts:
#	composer.lock

jmarx committed:
Fix composer.json/lock mismatch: add wp-ai-client to lock file

jmarx committed:
fix all php things and a bug i found with venues/events images'
'

jmarx committed:
PHPstan fixes

jmarx committed:
Fix javascript stuff
